# Call Center Cleanup

For this assignment, we will be working with call center data. You can start working on the assignment after the first lesson on Exploratory Data Analysis. Make sure to read the whole assignment before starting anything! As you code along in the Jupyter notebook, you are asked to make note of the results of your analysis. Do so by clicking on the results box and adding your notes beside each question.

## Business Issue and Understanding

You are working for a company that has two call centers: the North Call Center and the South Call Center. The company is looking to possibly hire five additional reps to enhance customer experience. Your task is to explore how efficient the current reps are in each branch to determine which branch would benefit from additional assistance.

### How the Call Center Works

Call center representatives are assigned queues. When calls are assigned to a queue, the call is assigned to the next person in line in the queue. After a call is assigned to a representative, the amount of time between assignment and the call starting is divided into busy minutes and not ready minutes. If the call is incoming and a customer is waiting on the phone for a rep, the time is split into three categories: busy minutes, not ready minutes, and incoming call wait time. Once the rep has the customer on the phone, there might be during call wait time, where the call is put on hold while the rep gets an answer for the customer.

### Notes about the Dataset

If you haven't worked in a call center before, these notes might help you throughout your analysis.

* The call purpose is tagged for each call.
* The time of the call is tagged in 1 hour blocks starting at 9:00 AM and ending at 5:00 PM.
* Calls are tagged as incoming or outgoing.
* Reps are assigned to queues. When the reps are working, they take calls in the order of their queue.
* A call that is dropped due to technical error or missed by the center because they have reached maximum capacity is a lost call.
* An abandoned call is when the customer hangs up because they have been waiting for too long.
* Busy Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are busy with other customers.
* Not Ready Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are not ready (for example, getting water).
* Incoming Wait Time - amount of time after assigned rep is available to take the call customer waits for representative to pick up a call. This is tracked in seconds.
* During Call Wait Time - amount of time during call that customer has to wait for representative

## Getting Started

You have two CSVs at your disposal, `NorthCallCenter.csv` and `SouthCallCenter.csv`. Import the appropriate libraries and create two dataframes, one called `north_df` and one called `south_df`.

In [31]:
# Import the appropriate libraries
import pandas as pd
pd.set_option("display.max_rows",None, "display.max_columns", None)


# Create two new dataframes
north_df = pd.read_csv("NorthCallCenter.csv")
north_df.tail(5)

south_df = pd.read_csv("SouthCallCenter.csv")
south_df = pd.set_option("display.max_rows",None, "display.max_columns", None)



In [50]:
south_df = pd.read_csv("SouthCallCenter.csv")
south_df.tail(313)
south_df['Time Block'].value_counts()

12:00 PM    75
10:00 AM    56
3:00 PM     43
11:00 AM    31
2:00 PM     30
5:00 PM     26
4:00 PM     25
1:00 PM     23
9:00 AM      5
Name: Time Block, dtype: int64

In [65]:
south_df.tail(313)
south_df['Calls'].value_counts()

1    314
Name: Calls, dtype: int64

In [66]:
south_df.describe( )

Unnamed: 0   Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  314.000000  314.000000  314.000000     314.00000         314.000000   
mean   156.500000    0.035032    0.012739      10.05414           1.910828   
std     90.788215    0.184154    0.112324       1.40751           0.860016   
min      0.000000    0.000000    0.000000       8.00000           1.000000   
25%     78.250000    0.000000    0.000000       9.00000           1.000000   
50%    156.500000    0.000000    0.000000      10.00000           2.000000   
75%    234.750000    0.000000    0.000000      11.00000           3.000000   
max    313.000000    1.000000    1.000000      12.00000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count          126.000000             314.000000  314.0  
mean             3.000000               3.076433    1.0  
std              1.308434               0.819424    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

## Exploratory Data Analysis

Time to do some EDA! In the process of learning more about the two datasets, make sure you answer the following questions:

1. How many reps are in each branch?
2. What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for each branch?
3. What is the number of calls taken for each time block?

In [64]:
# Put your EDA code here
north_df.tail(244)
north_df.describe()

north_df['Calls'].value_counts()

1    245
Name: Calls, dtype: int64

In [67]:
north_df.describe( )

Unnamed: 0   Lost Call   Abandoned  Busy Minutes  Not Ready Minutes  \
count  245.000000  245.000000  245.000000    245.000000         245.000000   
mean   122.000000    0.020408    0.028571      9.987755           1.914286   
std     70.869599    0.141681    0.166940      1.403980           0.771416   
min      0.000000    0.000000    0.000000      8.000000           1.000000   
25%     61.000000    0.000000    0.000000      9.000000           1.000000   
50%    122.000000    0.000000    0.000000     10.000000           2.000000   
75%    183.000000    0.000000    0.000000     11.000000           3.000000   
max    244.000000    1.000000    1.000000     12.000000           3.000000   

       Incoming Wait Time  During Call Wait Time  Calls  
count           82.000000             245.000000  245.0  
mean             3.048780               2.967347    1.0  
std              1.456382               0.843827    0.0  
min              1.000000               2.000000    1.0  
25%              2.000000               2.000000    1.0  
50%              3.000000               3.000000    1.0  
75%              4.000000               4.000000    1.0  
max              5.000000               4.000000    1.0

### EDA Results

Make note of your results here!

1. How many reps are in each branch? There are 244 reps in the North branch and 313 in the South.
2. What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for each branch? The North branch has an average of 9.987755 busy minutes, 1.914286 not ready minutes, 3.048780 minutes of incoming wait time, and 2.967347 minutes of during call wait time. The South branch  has an average of 10.05414 busy minutes, 1.910828 not ready minutes, 3.0 minutes of incoming wait time, and 3.076433 minutes of during call wait time. 
3. What is the number of calls taken for each time block? In the north branch there are 5 calls taken at 9:00am, 43 calls at 10:00am, 25 calls at 11:00 am, 45 calls taken at 12:00 pm, 17 calls taken at 1:00 pm, 35 calls taken at 2:00 pm, 30 calls taken at 3:00 pm,28 calls taken at 4:00pm, and 17 calss taken at 5 pm. At the south branch 5 calls were taken at 9:00am, 56 calls taken at 10:00 am,31 calls taken at 11:00 am, 75 calls at 12:00 pm, 23 at 1:00 pm, 30 at 2:00 pm, 43 at 3:00 pm, 25 at 4:00 pm,  and 26 at 5:00 pm. 

## Cleaning Data

Now you need to clean up the datasets. When cleaning the datasets, you may find that there isn't dirty data to clean. That is okay! Some questions you need to answer about the data sets:

1. Incoming wait time is null in many places. Do all of those places correspond to outgoing calls?
2. In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?
3. Are there duplicates in the two data sets? If there are, how did you handle them?
4. Is any of the data in the two data sets unnecessary? If yes, how did you handle it?

In [62]:
# Clean your data here
south_df.tail(313)


Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
1             1  South    Sales Support   10:00 AM             Incoming     C   
2             2  South    Sales Support   10:00 AM             Incoming     C   
3             3  South    Sales Support   10:00 AM             Incoming     C   
4             4  South    Sales Support   12:00 PM             Incoming     C   
5             5  South    Sales Support    1:00 PM             Incoming     D   
6             6  South  Product Support   11:00 AM             Incoming     C   
7             7  South  Product Support   12:00 PM             Incoming     C   
8             8  South  Product Support    2:00 PM             Incoming     D   
9             9  South  Product Support    3:00 PM             Incoming     D   
10           10  South  Product Support    3:00 PM             Incoming     D   
11           11  South  Product Support    4:00 PM             Incoming     D   
12           12  South  Product Support    4:00 PM             Incoming     D   
13           13  South  Product Support    5:00 PM             Incoming     C   
14           14  South  Product Support    5:00 PM             Incoming     D   
15           15  South        Complaint   11:00 AM             Incoming     D   
16           16  South        Complaint   12:00 PM             Incoming     D   
17           17  South        Complaint    1:00 PM             Incoming     C   
18           18  South        Complaint    1:00 PM             Incoming     C   
19           19  South        Complaint    5:00 PM             Incoming     D   
20           20  South        Complaint   12:00 PM             Incoming     D   
21           21  South    Sales Support   10:00 AM             Incoming     C   
22           22  South    Sales Support   10:00 AM             Incoming     D   
23           23  South    Sales Support   11:00 AM             Incoming     D   
24           24  South    Sales Support   12:00 PM             Incoming     C   
25           25  South    Sales Support    1:00 PM             Incoming     D   
26           26  South  Product Support   10:00 AM             Incoming     C   
27           27  South  Product Support   11:00 AM             Incoming     C   
28           28  South  Product Support   11:00 AM             Incoming     C   
29           29  South  Product Support   12:00 PM             Incoming     C   
30           30  South  Product Support    2:00 PM             Incoming     C   
31           31  South  Product Support    2:00 PM             Incoming     D   
32           32  South  Product Support    3:00 PM             Incoming     C   
33           33  South  Product Support    3:00 PM             Incoming     C   
34           34  South  Product Support    3:00 PM             Incoming     D   
35           35  South  Product Support    5:00 PM             Incoming     C   
36           36  South        Complaint    9:00 AM             Incoming     C   
37           37  South        Complaint   10:00 AM             Incoming     C   
38           38  South        Complaint   12:00 PM             Incoming     D   
39           39  South        Complaint   12:00 PM             Incoming     D   
40           40  South        Complaint   12:00 PM             Incoming     C   
41           41  South        Complaint    1:00 PM             Incoming     C   
42           42  South        Complaint    1:00 PM             Incoming     C   
43           43  South        Complaint    2:00 PM             Incoming     D   
44           44  South        Complaint    3:00 PM             Incoming     C   
45           45  South        Complaint    3:00 PM             Incoming     D   
46           46  South        Complaint    3:00 PM             Incoming     C   
47           47  South        Complaint    5:00 PM             Incoming     D   
48           48  South    Sales Support   10:00 AM             Incoming     D   
49           49  South    Sales Supp

In [60]:
north_df.tail(244)


Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
1             1  North    Sales Support   11:00 AM             Incoming     A   
2             2  North    Sales Support   11:00 AM             Incoming     B   
3             3  North    Sales Support    4:00 PM             Incoming     B   
4             4  North  Product Support   12:00 PM             Incoming     B   
5             5  North  Product Support    2:00 PM             Incoming     B   
6             6  North  Product Support    2:00 PM             Incoming     A   
7             7  North  Product Support    2:00 PM             Incoming     B   
8             8  North  Product Support    3:00 PM             Incoming     B   
9             9  North  Product Support    3:00 PM             Incoming     A   
10           10  North  Product Support    4:00 PM             Incoming     A   
11           11  North  Product Support    4:00 PM             Incoming     A   
12           12  North  Product Support    4:00 PM             Incoming     B   
13           13  North        Complaint    9:00 AM             Incoming     B   
14           14  North        Complaint   12:00 PM             Incoming     B   
15           15  North        Complaint    1:00 PM             Incoming     B   
16           16  North        Complaint    3:00 PM             Incoming     B   
17           17  North        Complaint    2:00 PM             Incoming     A   
18           18  North    Sales Support   10:00 AM             Incoming     B   
19           19  North    Sales Support   10:00 AM             Incoming     B   
20           20  North    Sales Support   11:00 AM             Incoming     B   
21           21  North    Sales Support   12:00 PM             Incoming     B   
22           22  North    Sales Support    1:00 PM             Incoming     A   
23           23  North    Sales Support    1:00 PM             Incoming     B   
24           24  North  Product Support   12:00 PM             Incoming     A   
25           25  North  Product Support    1:00 PM             Incoming     A   
26           26  North  Product Support    2:00 PM             Incoming     B   
27           27  North  Product Support    3:00 PM             Incoming     B   
28           28  North  Product Support    3:00 PM             Incoming     A   
29           29  North  Product Support    3:00 PM             Incoming     A   
30           30  North  Product Support    4:00 PM             Incoming     B   
31           31  North  Product Support    4:00 PM             Incoming     A   
32           32  North        Complaint   12:00 PM             Incoming     A   
33           33  North    Sales Support   10:00 AM             Incoming     B   
34           34  North    Sales Support   10:00 AM             Incoming     A   
35           35  North    Sales Support   10:00 AM             Incoming     B   
36           36  North    Sales Support   11:00 AM             Incoming     A   
37           37  North    Sales Support   11:00 AM             Incoming     A   
38           38  North    Sales Support   12:00 PM             Incoming     B   
39           39  North    Sales Support   12:00 PM             Incoming     B   
40           40  North    Sales Support   12:00 PM             Incoming     B   
41           41  North  Product Support    3:00 PM             Incoming     B   
42           42  North  Product Support    4:00 PM             Incoming     B   
43           43  North  Product Support    5:00 PM             Incoming     A   
44           44  North        Complaint    2:00 PM             Incoming     A   
45           45  North    Sales Support   10:00 AM             Incoming     B   
46           46  North    Sales Support   12:00 PM             Incoming     A   
47           47  North    Sales Support   10:00 AM             Incoming     A   
48           48  North    Sales Support   11:00 AM             Incoming     A   
49           49  North    Sales Supp

In [58]:
duplicates = north_df.duplicated( )
duplicates.sum()

0

### Cleaning Data Results

Make note of your results!

1. Incoming wait time is null in many places. Do all of those places correspond to outgoing calls? In both branches they all correspond to outgoing calls. 
2. In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?
3. Are there duplicates in the two data sets? If there are, how did you handle them? There are no duplicates in either data set.
4. Is any of the data in the two data sets unnecessary? If yes, how did you handle it? The queue isn't a necessary column. It could have been dropped but I did not. 

## Data Manipulation

Before you begin answering any questions, combine the two datasets together to create a third dataframe called `df`. You can use this third dataframe to compare the two call centers to company-wide trends.

1. Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.
892. The average call abandonment rate is a KPI when it comes tddddddddddddddddo determining call center efficiency. As you may recall, ab4444444444444444444444444444444444444444444400000000000000047andoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%? 
3. Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each branch and the entire company?
4. For each branch and the entire company, what is the average speed of answer?

In [63]:
# Manipulate data here
north_df.merge(south_df, on='Branch')

Empty DataFrame
Columns: [Unnamed: 0_x, Branch, Call Purpose_x, Time Block_x, Incoming or Outgoing_x, Queue_x, Rep ID_x, Sale_x, Lost Call_x, Abandoned_x, Busy Minutes_x, Not Ready Minutes_x, Incoming Wait Time_x, During Call Wait Time_x, Calls_x, Unnamed: 0_y, Call Purpose_y, Time Block_y, Incoming or Outgoing_y, Queue_y, Rep ID_y, Sale_y, Lost Call_y, Abandoned_y, Busy Minutes_y, Not Ready Minutes_y, Incoming Wait Time_y, During Call Wait Time_y, Calls_y]
Index: []

[0 rows x 29 columns]

### Data Manipulation Results

1. Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.
2. The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%? The North has a 0.02 average call abandonment rate and the South has 0.01. The company as a whole has an average of 0.03 call abandonment rate. 
3. Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each company and the entire company? 14.9% of calls are answered withing 2 seconds in the south branch and 13.0% in the North. 14% as a whole. 
4. For each branch and the entire company, what is the average speed of answer? 3.0% in the north and south. For the entire company it is 3.0%.

## Visualization

Create a visualization for each of the following questions. Some of the code to handle aggregating and storing data may be written for you. For each visualization, you choose the chart style that you feel suits the situation best. Make note of the chart style you chose and why.

1. What is the average abandonment rate per queue?
2. What is the service level and average speed of answer per each rep?
3. For each type of call purpose, how many calls are outgoing vs. incoming?

In [11]:
# Create visualization 1 here
# The dictionary abandonment_rates has the data you need.

abandonment_rates = {}
queues = ["A", "B", "C", "D"]
queue_dict = df.groupby("Queue").agg("sum")
for i in range(4):
    abandonment_rates[queues[i]] = queue_dict["Abandoned"][i] / queue_dict["Calls"][i] 

NameError: name 'df' is not defined

In [ ]:
# Create visualization 2 here
# north_plt contains the data you need for the average speed of answer of each rep

north_plt = north.groupby("Rep ID")["Incoming Wait Time"].mean().to_frame().reset_index()

# Finding each Rep's Personal Service Level Percentage.  Basically, Calls within 2 secs / total calls

# Table 1: Total Incoming calls less than 2 seconds grouped by Rep
quick_calls = north[north["Incoming Wait Time"] <= 2.0]
quick_reps = quick_calls[["Rep ID", "Calls"]]
quick_stats = quick_reps.groupby(["Rep ID"]).sum()  # Final Table


# Table 2: Total Incoming Calls Only grouped by Rep
total_calls_in = north[north["Incoming or Outgoing"] == "Incoming"]
rep_calls = total_calls_in[["Rep ID", "Calls"]]     
total_stats = rep_calls.groupby(["Rep ID"]).sum() # Final Table  

#  Table 3: Service Level Percentage created via merge
service_level = pd.merge(quick_stats, total_stats, on="Rep ID")

# Create Percentage Column in Table 3
service_level["Service Level %"] = service_level["Calls_x"]/service_level["Calls_y"] * 100

In [ ]:
# Create visualization 3 here
# The three dictionaries, complaints, sales_support, and product_support, have the information you need

purpose_counts = purpose_group["Incoming or Outgoing"].value_counts()
complaints = purpose_counts["Complaint"].to_dict()
sales_support = purpose_counts["Sales Support"].to_dict()
product_support = purpose_counts["Product Support"].to_dict()

### Visualization Results

For each chart you created, explain why you chose the chart style you chose.

## Summarize Your Work

With what you know now about the two call centers and the entire company, answer the following questions. Note that while this is subjective, you should include relevant data to back up your opinion.

1. Using KPIs such as average abandonment rate, service level and average speed of answer, in your opinion, which one of the two branches is operating more efficiently? Why? The South branch is working more efficiently because they have the lowest numbers in mentioned KPI's. 
2. Based on the number of reps in each branch and how quickly the reps are working, in your opinion, which branch would benefit from the extra help? The north could benefit from the extra help. 
3. Now that you have explored the datasets, is there any data or information that you wish you had in this analysis? There is not any information that I wish I had. 

## Submit Your Work

After you have completed your work on the assignment, push your work to your Github repo. Navigate to the repo, copy the link to your assignment and paste the URL in the Canvas assignment's submission box. 

## Bonus Mission

Create a visualization that answers this question:

1. For each call purpose, how many calls (incoming and outgoing) take place in each time block?

In [ ]:
# Create your visualization here!
call_times = df[["Time Block", "Call Purpose", "Incoming or Outgoing", "Calls"]]

# Use groupby to plot based on time blocks:

# Use groupby and get_group to select which call purpose to plot:
